### PR TITLE
cilium-cni: implement cni CHECK support

### LINKED ISSUE #17251: Support CNI v0.4.0 with new CHECK action
Currently, cilium-cni supports 0.1.x-0.3.x.

- Kubernetes supports CNI v.0.4.0
- Coil(https://github.com/cybozu-go/coil) uses only CNI v0.4.0, so it cannot be used in combination with cilium.
- By supporting v0.4.0, cilium will be able to use the CHECK command.

It would be nice if cIlium could support v0.4.0.

### PR DESCRIPTION
**Pull request by squeed**
Fixes: #17251

The cni CHECK action asks the plugin to ensure that the container's networking is configured as desired.

Fortunately, the agent already exposes a "healthz"-style api; all we need to do is call it. Also, verify that the veth interface exists and is configured correctly.

```release-note
The default CNI version is now v0.4.0. Cilium now supports the CNI CHECK action.
```

Note that, right now, very few runtimes actually call CNI CHECK. I tested this with a hacked version of containerd that called CNI CHECK periodically. Some of the other CNI maintainers are working on rolling out check support universally.

### CODE REVIEW

**Comment by tommyp1ckles** [comment]
@squeed Are we intending to change the cni version for aws-cni chaining to support newer versions of aws-vpc-cni?

File: plugins/cilium-cni/main.go
Comment (tommyp1ckles) [inline plugins/cilium-cni/main.go]:
```suggestion
	// If this is a chained plugin, token "delegate" to the special chaining mode and be done
```

**Comment by squeed** [comment]
> @squeed Are we intending to change the cni version for aws-cni chaining to support newer versions of aws-vpc-cni? 

Good point, all the CNI versions should be bumped.

File: plugins/cilium-cni/main.go
Comment (squeed) [reply plugins/cilium-cni/main.go]: should be "then" :-p

**Comment by squeed** [comment]
OK, rebased, fixed lint error, and bumped all CNI configs to v0.4.0. We might have an issue with very old flannel installations; hopefully this won't be an issue.

**Comment by squeed** [comment]
One last TODO: implement this for the other "chaining" modes.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@squeed do you wan to implement the chaining as part of this PR or in a follow up?

**Comment by squeed** [comment]
> @squeed do you wan to implement the chaining as part of this PR or in a follow up?

I'll take care of it in this PR; should be ready soon.

**Comment by squeed** [comment]
@aanm implemented chained check too. It was easy enough.

This is ready to go. Next TODO: get check in to ContainerD :-p

**Comment by aanm** [comment]
@squeed it looks the CI is complaining

File: plugins/cilium-cni/main.go
Comment (tommyp1ckles) [inline plugins/cilium-cni/main.go:833]: Question: Under what circumstance is this empty?

**Comment by tommyp1ckles** [comment]
Cool stuff, looks good to me aside from CI issues @squeed Is there going to be any immediate effects from this aside from CNI compatibility. I.e. Does Kubelet perform CHECKs periodically?

**Review by tommyp1ckles — APPROVED** [review]

File: plugins/cilium-cni/main.go
Comment (squeed) [reply plugins/cilium-cni/main.go:833]: Great question. the CNI Result type allows you to add non-sandbox interfaces if they are relevant (e.g. the bridge on the host). The way to signify this is that the Sandbox field is empty.

**Comment by squeed** [comment]
> Cool stuff, looks good to me aside from CI issues @squeed Is there going to be any immediate effects from this aside from CNI compatibility. I.e. Does Kubelet perform CHECKs periodically?

No,

**Comment by squeed** [comment]
> Cool stuff, looks good to me aside from CI issues @squeed Is there going to be any immediate effects from this aside from CNI compatibility. I.e. Does Kubelet perform CHECKs periodically?

Right now, runtime use of CHECK is somewhat limited. ContainerD doesn't call it yet (but I hope to change that soon). Likewise, cro-o only calls it when cri-o restarts.

So, right now, there should be essentially no impact (sadly). But I am actively working on changing that.

**Comment by squeed** [comment]
> @squeed it looks the CI is complaining

Whoops, fixed.

**Comment by squeed** [comment]
Interesting; that test failure is with the portmap plugin, which I theoretically touched (but didn't really). I'll look more in to it.

**Comment by squeed** [comment]
Given that the test passes on other kernel versions, and that I haven't touched anything interesting w.r.t. portmap, I'm judging the failure to be a flake.

/test

**Comment by pchaigno** [comment]
> Given that the test passes on other kernel versions, and that I haven't touched anything interesting w.r.t. portmap

Please note that the Portmap test only runs on that 4.9 CI job; it is skipped in other CI jobs. So if the test ends up being broken, this pull request is likely the cause. If it ends up being more flaky in the coming days, this pull request may be the cause.

**Comment by tklauser** [comment]
It looks like `K8sConformance Portmap Chaining Check connectivity-check compliance with portmap chaining` is failing consistently since Sep 1 on `master`: https://jenkins.cilium.io/job/cilium-master-k8s-1.16-kernel-4.9/

Given that the PR was merged on Sep 1, the test also failed on this PR and Paul's [comment](https://github.com/cilium/cilium/pull/20956#issuecomment-1236164709) above I'd assume this PR to be the likely culprit.

**Comment by joestringer** [comment]
> Given that the test passes on other kernel versions, and that I haven't touched anything interesting w.r.t. portmap, I'm judging the failure to be a flake.

Just a heads up, I would expect that every flake should have a corresponding github issue. This way, other contributors can find those flakes and corroborate that they are unrelated to your changes. If you are unable to find an issue corresponding to the actual failure, then that typically suggests the issue may be related to your PR. In that case, I would encourage raising the failure for discussion in the OSS Slack #testing or #development channels. (That said, obviously the flake must be first observed at some point so we do need to use our judgement to choose when to file the flake in the first place; for that, it may be useful to search the title of the test in the issues & PRs to see if you can locate the same failure elsewhere).

**Comment by pchaigno** [comment]
@joestringer Note it's not a flake but a complete breakage. For those, I would expect us to fix in a few hours and thus not need an issue (as long as we warn on Slack to rebase after the fix is merged). That said, my expectation doesn't often meet reality :disappointed:

**Comment by julianwiedmann** [comment]
For linkage purposes - the offending commit has been reverted with https://github.com/cilium/cilium/pull/21207.

**Comment by joestringer** [comment]
> @joestringer Note it's not a flake but a complete breakage. For those, I would expect us to fix in a few hours and thus not need an issue (as long as we warn on Slack to rebase after the fix is merged). That said, my expectation doesn't often meet reality disappointed

**Comment by joestringer** [comment]
> @joestringer Note it's not a flake but a complete breakage. For those, I would expect us to fix in a few hours and thus not need an issue (as long as we warn on Slack to rebase after the fix is merged). That said, my expectation doesn't often meet reality disappointed

I agree. My point was that prior to merging, if you believe something to be a flake, then you can search the issues to find out. It appears that we merged this PR under the assumption that the issue was a flake, but it was not actually a flake.

I'm not fussed about whether we file an issue or not for a real issue, filing an issue is a low-effort thing we can do to document the issue & centralize the tracking for it (and help others search for it). If we identify the issue quickly and get the author/maintainer in the loop for fixing the issue, then we don't necessarily have to file an issue; we can just collaborate on a revert or fix and then announce on slack when the issue is fixed.

**Pull request merged**
